Golang lint

```
Running golangci-lint: FAIL
level=warning msg="[runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive."
level=warning msg="[linters context] structcheck is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649."
pkg/metrics/metrics.go:35: File is not `gofmt`-ed with `-s` (gofmt)
	stackType					string
pkg/metrics/collector/pod_ip_metrics.go:70:2: var `duplicateIpCountDesc` should be `duplicateIPCountDesc` (golint)
	duplicateIpCountDesc = prometheus.NewDesc(
	^
pkg/metrics/collector/pod_ip_metrics.go:94:2: var `podIpMetricsWatcherSetup` should be `podIPMetricsWatcherSetup` (golint)
	podIpMetricsWatcherSetup = false
	^
pkg/metrics/collector/pod_ip_metrics.go:97:6: type `podIpMetricsCollector` should be `podIPMetricsCollector` (golint)
type podIpMetricsCollector struct {
     ^
pkg/metrics/collector/pod_ip_metrics.go:102:2: struct field `duplicateIpCount` should be `duplicateIPCount` (golint)
	duplicateIpCount    uint64
	^
pkg/metrics/collector/pod_ip_metrics.go:141:6: func NewPodIpMetricsCollector should be NewPodIPMetricsCollector (golint)
func NewPodIpMetricsCollector() (Collector, error) {
     ^
pkg/metrics/collector/pod_ip_metrics.go:162:33: method listIpAddresses should be listIPAddresses (golint)
func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
                                ^
pkg/metrics/collector/pod_ip_metrics.go:175:28: var `dupIpCount` should be `dupIPCount` (golint)
	var ipv4Count, ipv6Count, dupIpCount, dualCount, dualErrCount uint64
	                          ^
pkg/metrics/collector/pod_ip_metrics.go:[191](https://github.com/GoogleCloudPlatform/netd/actions/runs/3301658861/jobs/5447391570#step:3:192):3: var `podId` should be `podID` (golint)
		podId, err := readLine(fileName)
		^
pkg/metrics/collector/pod_ip_metrics.go:235:33: method updateReuseIpStats should be updateReuseIPStats (golint)
func (c *podIpMetricsCollector) updateReuseIpStats(e fsnotify.Event, f string) {
                                ^
pkg/metrics/collector/pod_ip_metrics.go:236:2: var `reuseIp` should be `reuseIP` (golint)
	reuseIp, ok := c.reuseMap[f]
	^
pkg/metrics/collector/pod_ip_metrics.go:250:5: should replace `c.reuseIps.count += 1` with `c.reuseIps.count++` (golint)
				c.reuseIps.count += 1
				^
pkg/metrics/collector/pod_ip_metrics_test.go:44:6: func mustCreateIpAddrDir should be mustCreateIPAddrDir (golint)
func mustCreateIpAddrDir(t *testing.T, dir string) string {
     ^
pkg/metrics/collector/pod_ip_metrics_test.go:87:3: struct field `wantDuplicateIpCount` should be `wantDuplicateIPCount` (golint)
		wantDuplicateIpCount  uint64
		^
pkg/metrics/collector/pod_ip_metrics.go:339:84: unnecessary conversion (unconvert)
	ch <- prometheus.MustNewConstMetric(ipReuseAvgDesc, prometheus.GaugeValue, float64((c.reuseIps.sum / float64(c.reuseIps.count))))
	                                                                                  ^
pkg/metrics/collector/pod_ip_metrics_test.go:22:2: S1023: redundant `return` statement (gosimple)
	return
	^
pkg/metrics/collector/helper.go:20:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
pkg/metrics/collector/kernel_metrics.go:21:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
pkg/metrics/collector/netlink_metrics.go:21:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
pkg/metrics/collector/pod_ip_metrics_test.go:4:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
```

--- a/.github/workflows/lint.yaml
+++ b/.github/workflows/lint.yaml
@@ -1,0 +1,12 @@
+name: Run lint
+
+on: [ push, pull_request ]
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
+    - name: Lint
+      run: make -f Makefile.all lint

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -13,4 +13,4 @@ jobs:
     - name: Checkout code
       uses: actions/checkout@v3
     - name: Test
-      run: make test GOFLAGS="-v" GO_VERSION=${{ matrix.go-version }}
+      run: make -f Makefile.all test GOFLAGS="-v" GO_VERSION=${{ matrix.go-version }}

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -59,24 +59,22 @@ output:
 linters:
   disable-all: true
   enable:
-  - deadcode
   - errcheck
   - gocritic
   - gofmt
   - goimports
-  - golint
   - gosimple
   - govet
   - ineffassign
   - lll
   - misspell
+  - revive
   - staticcheck
-  - structcheck
   - stylecheck
   - typecheck
   - unconvert
   - unparam
-  - varcheck
+  - unused
   fast: false
 
 linters-settings:

--- a/pkg/config/config_policy_routing.go
+++ b/pkg/config/config_policy_routing.go
@@ -62,7 +62,6 @@ var (
 	defaultGateway   net.IP
 	defaultLinkIndex int
 	defaultNetdev    string
-	localLinkIndex   int
 	localNetdev      string
 )
 
@@ -91,7 +90,7 @@ func init() {
 		return
 	}
 	defaultLinkIndex, defaultNetdev, defaultGateway = f(net.IPv4(8, 8, 8, 8))
-	localLinkIndex, localNetdev, _ = f(net.IPv4(127, 0, 0, 1))
+	_, localNetdev, _ = f(net.IPv4(127, 0, 0, 1))
 
 	sysctlReversePathFilter := fmt.Sprintf("net.ipv4.conf.%s.rp_filter", defaultNetdev)
 	hairpinMaskStr := fmt.Sprintf("0x%x", hairpinMask)

--- a/pkg/metrics/collector/helper.go
+++ b/pkg/metrics/collector/helper.go
@@ -17,7 +17,7 @@ limitations under the License.
 package collector
 
 import (
-	"io/ioutil"
+	"os"
 	"path"
 	"strconv"
 	"strings"
@@ -29,7 +29,7 @@ var stackType string
 // readUintFromFile reads a uint from file given by path. The file is supposed to only contain a single uint
 // in base 10. Otherwise, an error will be returned.
 func readUintFromFile(path string) (uint64, error) {
-	data, err := ioutil.ReadFile(path)
+	data, err := os.ReadFile(path)
 	if err != nil {
 		return 0, err
 	}

--- a/pkg/metrics/collector/kernel_metrics.go
+++ b/pkg/metrics/collector/kernel_metrics.go
@@ -18,7 +18,7 @@ package collector
 
 import (
 	"fmt"
-	"io/ioutil"
+	"os"
 	"strconv"
 	"strings"
 
@@ -116,7 +116,7 @@ func parseKeyValueLines(output, label string) (map[string]uint64, error) {
 
 func (c *kernelStatCollector) Update(ch chan<- prometheus.Metric) error {
 	// Get snmp values.
-	data, err := ioutil.ReadFile("/proc/net/snmp")
+	data, err := os.ReadFile("/proc/net/snmp")
 	if err != nil {
 		return fmt.Errorf("could not read proc/net/snmp")
 	}
@@ -136,7 +136,7 @@ func (c *kernelStatCollector) Update(ch chan<- prometheus.Metric) error {
 	}
 
 	// Get netstat values
-	data, err = ioutil.ReadFile("/proc/net/netstat")
+	data, err = os.ReadFile("/proc/net/netstat")
 	if err != nil {
 		return fmt.Errorf("could not read proc/net/netstat")
 	}

--- a/pkg/metrics/collector/netlink_metrics.go
+++ b/pkg/metrics/collector/netlink_metrics.go
@@ -18,7 +18,6 @@ package collector
 import (
 	"flag"
 	"fmt"
-	"io/ioutil"
 	"os"
 	"strconv"
 	"sync"
@@ -235,7 +234,7 @@ func NewNetlinkCollector() (Collector, error) {
 }
 
 func getNamespaces() ([]netns.NsHandle, error) {
-	files, err := ioutil.ReadDir("/proc")
+	files, err := os.ReadDir("/proc")
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/metrics/collector/pod_ip_metrics.go
+++ b/pkg/metrics/collector/pod_ip_metrics.go
@@ -47,12 +47,12 @@ const (
 )
 
 var (
-	usedIpv4AddrCountDesc = prometheus.NewDesc(
+	usedIPv4AddrCountDesc = prometheus.NewDesc(
 		"used_ipv4_addr_count",
 		"Indicates how many IPv4 addresses are in use.",
 		nil, nil,
 	)
-	usedIpv6AddrCountDesc = prometheus.NewDesc(
+	usedIPv6AddrCountDesc = prometheus.NewDesc(
 		"used_ipv6_addr_count",
 		"Indicates how many IPv6 addresses are in use.",
 		nil, nil,
@@ -67,7 +67,7 @@ var (
 		"Indicates how many pods did not get dual stack IPs erroneously.",
 		nil, nil,
 	)
-	duplicateIpCountDesc = prometheus.NewDesc(
+	duplicateIPCountDesc = prometheus.NewDesc(
 		"duplicate_ip_count",
 		"Indicates how many pods had more than one IP per family.",
 		nil, nil,
@@ -91,21 +91,21 @@ var (
 		"Indicates the IP reuse duration in millisecond for all IPs.",
 		nil, nil,
 	)
-	podIpMetricsWatcherSetup = false
+	podIPMetricsWatcherSetup = false
 )
 
-type podIpMetricsCollector struct {
-	usedIpv4AddrCount   uint64
-	usedIpv6AddrCount   uint64
+type podIPMetricsCollector struct {
+	usedIPv4AddrCount   uint64
+	usedIPv6AddrCount   uint64
 	dualStackCount      uint64
 	dualStackErrorCount uint64
-	duplicateIpCount    uint64
-	reuseIps            reuseIps
+	duplicateIPCount    uint64
+	reuseIPs            reuseIPs
 	reuseMap            map[string]*ipReuse
 	clock               clock
 }
 
-type reuseIps struct {
+type reuseIPs struct {
 	count   uint64
 	sum     float64
 	min     uint64
@@ -133,13 +133,13 @@ func (ip ipFamily) String() string {
 }
 
 func init() {
-	registerCollector("pod_ip_metrics", NewPodIpMetricsCollector)
+	registerCollector("pod_ip_metrics", NewPodIPMetricsCollector)
 }
 
 // NewPodIpMetricsCollector returns a new Collector exposing pod IP allocation
 // stats.
-func NewPodIpMetricsCollector() (Collector, error) {
-	return &podIpMetricsCollector{clock: &realClock{}}, nil
+func NewPodIPMetricsCollector() (Collector, error) {
+	return &podIPMetricsCollector{clock: &realClock{}}, nil
 }
 
 func readLine(path string) (string, error) {
@@ -159,7 +159,7 @@ func readLine(path string) (string, error) {
 	return s.Text(), s.Err()
 }
 
-func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
+func (c *podIPMetricsCollector) listIPAddresses(dir string) error {
 	f, err := os.Open(dir)
 	if err != nil {
 		glog.Errorf("Error opening directory %s, %v", dir, err)
@@ -172,7 +172,7 @@ func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
 	}
 
 	podMap := make(map[string]ipFamily)
-	var ipv4Count, ipv6Count, dupIpCount, dualCount, dualErrCount uint64
+	var ipv4Count, ipv6Count, dupIPCount, dualCount, dualErrCount uint64
 	for _, v := range files {
 		if ip := net.ParseIP(v.Name()); ip == nil {
 			// This isn't an IP address file
@@ -188,18 +188,18 @@ func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
 		}
 		// Update the map and track IP families only for dual stack clusters
 		fileName := fmt.Sprintf("%s/%s", dir, v.Name())
-		podId, err := readLine(fileName)
+		podID, err := readLine(fileName)
 		if err != nil {
 			glog.Errorf("Error reading file %s: %v", fileName, err)
 			continue
 		}
-		f, ok := podMap[podId]
+		f, ok := podMap[podID]
 		if !ok {
-			podMap[podId] = family
+			podMap[podID] = family
 		} else if (f == ipv4 && family == ipv6) || (f == ipv6 && family == ipv4) {
-			podMap[podId] = dual
+			podMap[podID] = dual
 		} else {
-			dupIpCount++
+			dupIPCount++
 		}
 	}
 	if stackType == dualStack {
@@ -211,11 +211,11 @@ func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
 			}
 		}
 	}
-	c.usedIpv4AddrCount = ipv4Count
-	c.usedIpv6AddrCount = ipv6Count
+	c.usedIPv4AddrCount = ipv4Count
+	c.usedIPv6AddrCount = ipv6Count
 	c.dualStackCount = dualCount
 	c.dualStackErrorCount = dualErrCount
-	c.duplicateIpCount = dupIpCount
+	c.duplicateIPCount = dupIPCount
 	return nil
 }
 
@@ -232,25 +232,25 @@ func getIPv4(s string) (bool, string) {
 // After the ip is removed, it will be put into the reuseMap.
 // After the ip is reused, the reuseMap will update the ip reuse interval.
 // The reuseMap will not record any ip that is being used unless it is reused.
-func (c *podIpMetricsCollector) updateReuseIpStats(e fsnotify.Event, f string) {
-	reuseIp, ok := c.reuseMap[f]
+func (c *podIPMetricsCollector) updateReuseIPStats(e fsnotify.Event, f string) {
+	reuseIP, ok := c.reuseMap[f]
 	switch {
 	case e.Op&fsnotify.Remove == fsnotify.Remove:
 		if !ok {
 			c.reuseMap[f] = &ipReuse{c.clock.Now(), 0}
 		} else {
-			reuseIp.ipReleasedTimestamp = c.clock.Now()
+			reuseIP.ipReleasedTimestamp = c.clock.Now()
 		}
 	case e.Op&fsnotify.Create == fsnotify.Create:
 		if ok {
-			oldT := reuseIp.ipReleasedTimestamp
+			oldT := reuseIP.ipReleasedTimestamp
 			diff := uint64(c.clock.Now().Sub(oldT).Milliseconds())
 			if diff > 0 {
-				reuseIp.ipReuseInterval = float64(diff)
-				c.reuseIps.count += 1
-				c.reuseIps.sum += float64(diff)
-				if diff < c.reuseIps.min {
-					c.reuseIps.min = diff
+				reuseIP.ipReuseInterval = float64(diff)
+				c.reuseIPs.count++
+				c.reuseIPs.sum += float64(diff)
+				if diff < c.reuseIPs.min {
+					c.reuseIPs.min = diff
 				}
 				c.fillBuckets(diff)
 			}
@@ -258,16 +258,16 @@ func (c *podIpMetricsCollector) updateReuseIpStats(e fsnotify.Event, f string) {
 	}
 }
 
-func (c *podIpMetricsCollector) fillBuckets(diff uint64) {
+func (c *podIPMetricsCollector) fillBuckets(diff uint64) {
 	for _, bound := range bucketKeys {
 		if float64(diff) < bound {
-			c.reuseIps.buckets[bound]++
+			c.reuseIPs.buckets[bound]++
 		}
 	}
 }
 
-func (c *podIpMetricsCollector) setupDirectoryWatcher(dir string) error {
-	if err := c.listIpAddresses(dir); err != nil {
+func (c *podIPMetricsCollector) setupDirectoryWatcher(dir string) error {
+	if err := c.listIPAddresses(dir); err != nil {
 		return err
 	}
 	watcher, err := fsnotify.NewWatcher()
@@ -276,17 +276,17 @@ func (c *podIpMetricsCollector) setupDirectoryWatcher(dir string) error {
 		return err
 	}
 
-	c.reuseIps.min = uint64(math.Inf(+1))
-	c.reuseIps.buckets = make(map[float64]uint64)
+	c.reuseIPs.min = uint64(math.Inf(+1))
+	c.reuseIPs.buckets = make(map[float64]uint64)
 	for _, bound := range bucketKeys {
-		c.reuseIps.buckets[bound] = 0
+		c.reuseIPs.buckets[bound] = 0
 	}
 	c.reuseMap = make(map[string]*ipReuse)
 
 	go func() {
 		defer func() {
 			watcher.Close()
-			podIpMetricsWatcherSetup = false
+			podIPMetricsWatcherSetup = false
 		}()
 
 		for {
@@ -296,13 +296,13 @@ func (c *podIpMetricsCollector) setupDirectoryWatcher(dir string) error {
 					glog.Error("watcher is not ok")
 					return
 				}
-				if err := c.listIpAddresses(dir); err != nil {
+				if err := c.listIPAddresses(dir); err != nil {
 					return
 				}
 				// Only update the ip reuse mininum, average and histogram for IPv4.
 				ok, f := getIPv4(e.Name)
 				if ok {
-					c.updateReuseIpStats(e, f)
+					c.updateReuseIPStats(e, f)
 				}
 
 			case err, ok := <-watcher.Errors:
@@ -318,26 +318,26 @@ func (c *podIpMetricsCollector) setupDirectoryWatcher(dir string) error {
 	if err != nil {
 		glog.Errorf("Failed to add watcher for directory %s: %v", dir, err)
 	}
-	podIpMetricsWatcherSetup = true
+	podIPMetricsWatcherSetup = true
 	return nil
 }
 
-func (c *podIpMetricsCollector) Update(ch chan<- prometheus.Metric) error {
-	if !podIpMetricsWatcherSetup {
+func (c *podIPMetricsCollector) Update(ch chan<- prometheus.Metric) error {
+	if !podIPMetricsWatcherSetup {
 		if err := c.setupDirectoryWatcher(gkePodNetworkDir); err != nil {
 			glog.Errorf("setupDirectoryWatcher returned error: %v", err)
 			return nil
 		}
 	}
-	ch <- prometheus.MustNewConstMetric(usedIpv4AddrCountDesc, prometheus.GaugeValue, float64(c.usedIpv4AddrCount))
-	ch <- prometheus.MustNewConstMetric(usedIpv6AddrCountDesc, prometheus.GaugeValue, float64(c.usedIpv6AddrCount))
+	ch <- prometheus.MustNewConstMetric(usedIPv4AddrCountDesc, prometheus.GaugeValue, float64(c.usedIPv4AddrCount))
+	ch <- prometheus.MustNewConstMetric(usedIPv6AddrCountDesc, prometheus.GaugeValue, float64(c.usedIPv6AddrCount))
 	ch <- prometheus.MustNewConstMetric(dualStackCountDesc, prometheus.GaugeValue, float64(c.dualStackCount))
 	ch <- prometheus.MustNewConstMetric(dualStackErrorCountDesc, prometheus.GaugeValue, float64(c.dualStackErrorCount))
-	ch <- prometheus.MustNewConstMetric(duplicateIpCountDesc, prometheus.GaugeValue, float64(c.duplicateIpCount))
+	ch <- prometheus.MustNewConstMetric(duplicateIPCountDesc, prometheus.GaugeValue, float64(c.duplicateIPCount))
 
-	ch <- prometheus.MustNewConstMetric(ipReuseMinDesc, prometheus.GaugeValue, float64(c.reuseIps.min))
-	ch <- prometheus.MustNewConstMetric(ipReuseAvgDesc, prometheus.GaugeValue, float64((c.reuseIps.sum / float64(c.reuseIps.count))))
-	ch <- prometheus.MustNewConstHistogram(ipReuseHistogramDesc, c.reuseIps.count, c.reuseIps.sum, c.reuseIps.buckets)
+	ch <- prometheus.MustNewConstMetric(ipReuseMinDesc, prometheus.GaugeValue, float64(c.reuseIPs.min))
+	ch <- prometheus.MustNewConstMetric(ipReuseAvgDesc, prometheus.GaugeValue, c.reuseIPs.sum/float64(c.reuseIPs.count))
+	ch <- prometheus.MustNewConstHistogram(ipReuseHistogramDesc, c.reuseIPs.count, c.reuseIPs.sum, c.reuseIPs.buckets)
 
 	return nil
 }

--- a/pkg/metrics/collector/pod_ip_metrics_test.go
+++ b/pkg/metrics/collector/pod_ip_metrics_test.go
@@ -1,7 +1,6 @@
 package collector
 
 import (
-	"io/ioutil"
 	"os"
 	"testing"
 	"time"
@@ -19,7 +18,6 @@ func (c *fakeClock) Now() time.Time {
 
 func (c *fakeClock) Sleep(d time.Duration) {
 	c.now = c.now.Add(d)
-	return
 }
 
 func mustCreateFile(t *testing.T, dir, name, content string) {
@@ -41,8 +39,8 @@ func mustDeleteFile(t *testing.T, dir, name string) {
 	}
 }
 
-func mustCreateIpAddrDir(t *testing.T, dir string) string {
-	tempDir, err := ioutil.TempDir("", dir)
+func mustCreateIPAddrDir(t *testing.T, dir string) string {
+	tempDir, err := os.MkdirTemp("", dir)
 	if err != nil {
 		t.Fatalf("Error creating dir %s: %v", dir, err)
 	}
@@ -55,7 +53,7 @@ func mustCreateIpAddrDir(t *testing.T, dir string) string {
 
 func TestListIpAddresses(t *testing.T) {
 	// dir1 has 2 IPv4 addresses and 2 IPv6 addresses for dual stack pods
-	dir1 := mustCreateIpAddrDir(t, "dir1")
+	dir1 := mustCreateIPAddrDir(t, "dir1")
 	mustCreateFile(t, dir1, "10.0.0.1", "hash1")
 	mustCreateFile(t, dir1, "10.0.0.2", "hash2")
 	mustCreateFile(t, dir1, "2600:1900::1", "hash1")
@@ -63,14 +61,14 @@ func TestListIpAddresses(t *testing.T) {
 
 	// dir2 has 2 IPv4 addresses and 2 IPv6 addresses but only one real dual stack
 	// pod
-	dir2 := mustCreateIpAddrDir(t, "dir2")
+	dir2 := mustCreateIPAddrDir(t, "dir2")
 	mustCreateFile(t, dir2, "10.0.0.1", "hash1")
 	mustCreateFile(t, dir2, "10.0.0.2", "hash2")
 	mustCreateFile(t, dir2, "2600:1900::1", "hash1")
 	mustCreateFile(t, dir2, "2600:1900::3", "hash3")
 
 	// dir3 has 2 IPv4 and IPv6 addresses allocated for the same pod
-	dir3 := mustCreateIpAddrDir(t, "dir3")
+	dir3 := mustCreateIPAddrDir(t, "dir3")
 	mustCreateFile(t, dir3, "10.0.0.1", "hash1")
 	mustCreateFile(t, dir3, "10.0.0.2", "hash1")
 	mustCreateFile(t, dir3, "2600:1900::1", "hash1")
@@ -80,11 +78,11 @@ func TestListIpAddresses(t *testing.T) {
 		desc                  string
 		dir                   string
 		stackType             string
-		wantUsedIpv4Count     uint64
-		wantUsedIpv6Count     uint64
+		wantUsedIPv4Count     uint64
+		wantUsedIPv6Count     uint64
 		wantDualStackCount    uint64
 		wantDualStackErrCount uint64
-		wantDuplicateIpCount  uint64
+		wantDuplicateIPCount  uint64
 		wantErr               bool
 	}{
 		{
@@ -97,23 +95,23 @@ func TestListIpAddresses(t *testing.T) {
 			desc:              "dir1 with 2 IPv4 and 2 IPv6 addresses. Single stack",
 			dir:               dir1,
 			stackType:         "IPV4",
-			wantUsedIpv4Count: 2,
-			wantUsedIpv6Count: 2,
+			wantUsedIPv4Count: 2,
+			wantUsedIPv6Count: 2,
 		},
 		{
 			desc:               "dir1 with 2 IPv4 and 2 IPv6 addresses. dual stack",
 			dir:                dir1,
 			stackType:          "IPV4_IPV6",
-			wantUsedIpv4Count:  2,
-			wantUsedIpv6Count:  2,
+			wantUsedIPv4Count:  2,
+			wantUsedIPv6Count:  2,
 			wantDualStackCount: 2,
 		},
 		{
 			desc:                  "dir2 with dual stack errors",
 			dir:                   dir2,
 			stackType:             "IPV4_IPV6",
-			wantUsedIpv4Count:     2,
-			wantUsedIpv6Count:     2,
+			wantUsedIPv4Count:     2,
+			wantUsedIPv6Count:     2,
 			wantDualStackCount:    1,
 			wantDualStackErrCount: 2,
 		},
@@ -121,25 +119,25 @@ func TestListIpAddresses(t *testing.T) {
 			desc:                 "dir2 with dual stack and dup IPs",
 			dir:                  dir3,
 			stackType:            "IPV4_IPV6",
-			wantUsedIpv4Count:    2,
-			wantUsedIpv6Count:    2,
+			wantUsedIPv4Count:    2,
+			wantUsedIPv6Count:    2,
 			wantDualStackCount:   1,
-			wantDuplicateIpCount: 2,
+			wantDuplicateIPCount: 2,
 		},
 	} {
 		t.Run(tc.desc, func(t *testing.T) {
 			stackType = tc.stackType
-			mc := podIpMetricsCollector{}
-			err := mc.listIpAddresses(tc.dir)
+			mc := podIPMetricsCollector{}
+			err := mc.listIPAddresses(tc.dir)
 			if gotErr := err != nil; gotErr != tc.wantErr {
 				t.Errorf("Expected error %t, got error: %v", tc.wantErr, err)
 				return
 			}
-			if mc.usedIpv4AddrCount != tc.wantUsedIpv4Count {
-				t.Errorf("usedIpv4AddrCount. want: %d, got %d", tc.wantUsedIpv4Count, mc.usedIpv4AddrCount)
+			if mc.usedIPv4AddrCount != tc.wantUsedIPv4Count {
+				t.Errorf("usedIpv4AddrCount. want: %d, got %d", tc.wantUsedIPv4Count, mc.usedIPv4AddrCount)
 			}
-			if mc.usedIpv6AddrCount != tc.wantUsedIpv6Count {
-				t.Errorf("usedIpv6AddrCount. want: %d, got %d", tc.wantUsedIpv6Count, mc.usedIpv6AddrCount)
+			if mc.usedIPv6AddrCount != tc.wantUsedIPv6Count {
+				t.Errorf("usedIpv6AddrCount. want: %d, got %d", tc.wantUsedIPv6Count, mc.usedIPv6AddrCount)
 			}
 			if mc.dualStackCount != tc.wantDualStackCount {
 				t.Errorf("dualStackCount. want: %d, got %d", tc.wantDualStackCount, mc.dualStackCount)
@@ -147,8 +145,8 @@ func TestListIpAddresses(t *testing.T) {
 			if mc.dualStackErrorCount != tc.wantDualStackErrCount {
 				t.Errorf("dualStackErrorCount. want: %d, got %d", tc.wantDualStackErrCount, mc.dualStackErrorCount)
 			}
-			if mc.duplicateIpCount != tc.wantDuplicateIpCount {
-				t.Errorf("duplicateIpCount. want: %d, got %d", tc.wantDuplicateIpCount, mc.duplicateIpCount)
+			if mc.duplicateIPCount != tc.wantDuplicateIPCount {
+				t.Errorf("duplicateIpCount. want: %d, got %d", tc.wantDuplicateIPCount, mc.duplicateIPCount)
 			}
 		})
 	}
@@ -156,7 +154,7 @@ func TestListIpAddresses(t *testing.T) {
 
 func TestSetupDirectoryWatcher(t *testing.T) {
 	// dir1 has 2 IPv4 addresses and 2 IPv6 addresses for dual stack pods
-	dir1 := mustCreateIpAddrDir(t, "dir1")
+	dir1 := mustCreateIPAddrDir(t, "dir1")
 	mustCreateFile(t, dir1, "10.0.0.1", "hash1")
 	mustCreateFile(t, dir1, "10.0.0.2", "hash2")
 	mustCreateFile(t, dir1, "2600:1900::1", "hash1")
@@ -165,16 +163,16 @@ func TestSetupDirectoryWatcher(t *testing.T) {
 
 	// Setup directory watcher and verify metrics
 	fakeClock := &fakeClock{}
-	mc := podIpMetricsCollector{clock: fakeClock}
+	mc := podIPMetricsCollector{clock: fakeClock}
 	bucketKeys := prometheus.LinearBuckets(5e3, 5e3, 12)
 	if err := mc.setupDirectoryWatcher(dir1); err != nil {
 		t.Fatalf("Got error %v while setting up directory watcher", err)
 	}
-	if mc.usedIpv4AddrCount != 2 {
-		t.Errorf("usedIpv4AddrCount. want: 2, got %d", mc.usedIpv4AddrCount)
+	if mc.usedIPv4AddrCount != 2 {
+		t.Errorf("usedIpv4AddrCount. want: 2, got %d", mc.usedIPv4AddrCount)
 	}
-	if mc.usedIpv6AddrCount != 2 {
-		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIpv6AddrCount)
+	if mc.usedIPv6AddrCount != 2 {
+		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIPv6AddrCount)
 	}
 	if mc.dualStackCount != 2 {
 		t.Errorf("dualStackCount. want: 2, got %d", mc.dualStackCount)
@@ -182,29 +180,29 @@ func TestSetupDirectoryWatcher(t *testing.T) {
 	if mc.dualStackErrorCount != 0 {
 		t.Errorf("dualStackErrorCount. want: 0, got %d", mc.dualStackErrorCount)
 	}
-	if mc.duplicateIpCount != 0 {
-		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIpCount)
+	if mc.duplicateIPCount != 0 {
+		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIPCount)
 	}
 	for _, bound := range bucketKeys {
-		v, ok := mc.reuseIps.buckets[bound]
+		v, ok := mc.reuseIPs.buckets[bound]
 		if !ok {
 			t.Errorf("buckets are initialized with 0 values. want: ok, got %v", ok)
 		} else if v != 0 {
 			t.Errorf("no file is deleted. want value to be 0, got %v", v)
 		}
 	}
-	if !podIpMetricsWatcherSetup {
+	if !podIPMetricsWatcherSetup {
 		t.Fatal("podIpMetricsWatcherSetup: want: true, got: false")
 	}
 
 	//Add a new file to the directory. Verify metrics
 	mustCreateFile(t, dir1, "10.0.0.3", "hash3")
 	time.Sleep(1 * time.Second)
-	if mc.usedIpv4AddrCount != 3 {
-		t.Errorf("usedIpv4AddrCount. want: 3, got %d", mc.usedIpv4AddrCount)
+	if mc.usedIPv4AddrCount != 3 {
+		t.Errorf("usedIpv4AddrCount. want: 3, got %d", mc.usedIPv4AddrCount)
 	}
-	if mc.usedIpv6AddrCount != 2 {
-		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIpv6AddrCount)
+	if mc.usedIPv6AddrCount != 2 {
+		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIPv6AddrCount)
 	}
 	if mc.dualStackCount != 2 {
 		t.Errorf("dualStackCount. want: 2, got %d", mc.dualStackCount)
@@ -212,18 +210,18 @@ func TestSetupDirectoryWatcher(t *testing.T) {
 	if mc.dualStackErrorCount != 1 {
 		t.Errorf("dualStackErrorCount. want: 1, got %d", mc.dualStackErrorCount)
 	}
-	if mc.duplicateIpCount != 0 {
-		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIpCount)
+	if mc.duplicateIPCount != 0 {
+		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIPCount)
 	}
 
 	//Remove a file from the directory. Verify metrics
 	mustDeleteFile(t, dir1, "10.0.0.3")
 	time.Sleep(1 * time.Second)
-	if mc.usedIpv4AddrCount != 2 {
-		t.Errorf("usedIpv4AddrCount. want: 2, got %d", mc.usedIpv4AddrCount)
+	if mc.usedIPv4AddrCount != 2 {
+		t.Errorf("usedIpv4AddrCount. want: 2, got %d", mc.usedIPv4AddrCount)
 	}
-	if mc.usedIpv6AddrCount != 2 {
-		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIpv6AddrCount)
+	if mc.usedIPv6AddrCount != 2 {
+		t.Errorf("usedIpv6AddrCount. want: 2, got %d", mc.usedIPv6AddrCount)
 	}
 	if mc.dualStackCount != 2 {
 		t.Errorf("dualStackCount. want: 2, got %d", mc.dualStackCount)
@@ -231,8 +229,8 @@ func TestSetupDirectoryWatcher(t *testing.T) {
 	if mc.dualStackErrorCount != 0 {
 		t.Errorf("dualStackErrorCount. want: 0, got %d", mc.dualStackErrorCount)
 	}
-	if mc.duplicateIpCount != 0 {
-		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIpCount)
+	if mc.duplicateIPCount != 0 {
+		t.Errorf("duplicateIpCount. want: 0, got %d", mc.duplicateIPCount)
 	}
 
 	// Remove and recreate the file. Verify metrics
@@ -250,7 +248,7 @@ func TestSetupDirectoryWatcher(t *testing.T) {
 	time.Sleep(1 * time.Second)
 
 	for _, bound := range bucketKeys {
-		v, ok := mc.reuseIps.buckets[bound]
+		v, ok := mc.reuseIPs.buckets[bound]
 		if !ok {
 			t.Errorf("reused ip: buckets are initialized with 0 values. want: ok, got %v", ok)
 		} else if bound == 5e3 && v != 0 {

--- a/pkg/metrics/metrics.go
+++ b/pkg/metrics/metrics.go
@@ -32,7 +32,7 @@ var mcfg struct {
 	enabledCollectors string
 	listenAddress     string
 	procPath          string
-	stackType					string
+	stackType         string
 }
 
 func init() {

--- a/scripts/Dockerfile.in
+++ b/scripts/Dockerfile.in
@@ -114,30 +114,13 @@ ARG HASH_BINARY=0
 COPY bin/{ARG_OS}_{ARG_ARCH}/{ARG_BIN} /{ARG_BIN}
 
 
-## Start building the final image. Using an intermediate so we can stop in the middle and run the tests.
+## Start building the final image.
 
-FROM {ARG_FROM} AS intermediate
-
-COPY --from=stuff / /
-
-
-## Run shell-test.sh in a standalone container. The outcome is discarded.
-
-FROM intermediate AS test
-
-COPY shell-test.sh /shell-test.sh
-
-RUN /shell-test.sh
-
-
-## Continue from the intermediate container (before running shell-test.sh) and complete the final image.
-
-FROM intermediate
+FROM {ARG_FROM}
 
 MAINTAINER Jingyuan Liang <jingyuanliang@google.com>
 
-# This file is not being used. Include it to trigger the test run and allow manual testing.
-COPY --from=test /shell-test.sh /shell-test.sh
+COPY --from=stuff / /
 
 ENV HOME /
 

--- a/scripts/Makefile
+++ b/scripts/Makefile
@@ -344,6 +344,51 @@ version: # @HELP outputs the version string
 version:
 	echo $(VERSION)
 
+test: # @HELP runs tests
+test: container
+	echo "Building and testing inotify"
+	docker run                              \
+	    -i                                  \
+	    --rm                                \
+	    -v $$(pwd)/inotify:/src             \
+	    --env HTTP_PROXY="$(HTTP_PROXY)"    \
+	    --env HTTPS_PROXY="$(HTTPS_PROXY)"  \
+	    alpine:3.16                         \
+	    sh -c 'cp -a /src /inotify && cd /inotify && ./build.sh && ./test.sh'
+	echo "Running shell-test.sh"
+	docker run                              \
+	    -i                                  \
+	    --rm                                \
+	    -v $$(pwd)/shell-test.sh:/test.sh   \
+	    --env HTTP_PROXY="$(HTTP_PROXY)"    \
+	    --env HTTPS_PROXY="$(HTTPS_PROXY)"  \
+	    $(REGISTRY)/$(BINS):$(TAG)          \
+	    /test.sh
+
+lint: # @HELP runs lint
+lint:
+	echo "Linting inotify"
+	docker run                              \
+	    -i                                  \
+	    --rm                                \
+	    -v $$(pwd)/inotify:/src             \
+	    -w /src                             \
+	    --env HTTP_PROXY="$(HTTP_PROXY)"    \
+	    --env HTTPS_PROXY="$(HTTPS_PROXY)"  \
+	    python:3-alpine                     \
+	    sh -c 'pip install cpplint && cpplint --recursive .'
+	echo "Linting install-cni.sh"
+	docker run                              \
+	    -i                                  \
+	    --rm                                \
+	    -u $$(id -u):$$(id -g)              \
+	    -v $$(pwd):/src                     \
+	    -w /src                             \
+	    --env HTTP_PROXY="$(HTTP_PROXY)"    \
+	    --env HTTPS_PROXY="$(HTTPS_PROXY)"  \
+	    koalaman/shellcheck                 \
+	    install-cni.sh
+
 $(BUILD_DIRS):
 	mkdir -p $@
 

--- a/scripts/inotify/build.sh
+++ b/scripts/inotify/build.sh
@@ -19,7 +19,9 @@ file inotify
 # exit with error code 1 if the executable is dynamic, not static
 ldd inotify && exit 1 || true
 
-./test.sh
+# quick test; full test in ./test.sh
+./inotify / '' /bin/true
 
 mkdir -p /tmp/release/
-mv inotify /tmp/release/inotify
+# cp instead of mv so we can run ./test.sh later
+cp inotify /tmp/release/inotify

--- a/scripts/inotify/inotify.c
+++ b/scripts/inotify/inotify.c
@@ -1,3 +1,5 @@
+// Copyright 2022 Google LLC
+
 #include <errno.h>
 #include <limits.h>
 #include <poll.h>
@@ -20,12 +22,12 @@ const int kPollTimeout = 2000;
 char buf[1024 * (sizeof(struct inotify_event) + NAME_MAX + 1)];
 
 void run_callback_internal(char** callback) {
-  fflush(stdout); // avoid mixing with callback output.
+  fflush(stdout);  // avoid mixing with callback output.
 
   pid_t pid = fork();
   if (pid == -1) {
     perror("fork");
-    return; // let it retry at the next event or timeout.
+    return;  // let it retry at the next event or timeout.
   }
 
   if (pid == 0) {
@@ -42,14 +44,16 @@ void run_callback_internal(char** callback) {
         continue;
       }
       perror("waitpid");
-      exit(EXIT_FAILURE); // shouldn't happen; exit & don't risk leaking zombies.
+      exit(EXIT_FAILURE);  // shouldn't happen; don't risk leaking zombies.
     }
   } while (!WIFEXITED(status) && !WIFSIGNALED(status));
 
   if (WIFEXITED(status)) {
-    printf("inotify: %s exited with status %d (exit: %d)\n", callback[0], status, WEXITSTATUS(status));
+    printf("inotify: %s exited with status %d (exit: %d)\n",
+           callback[0], status, WEXITSTATUS(status));
   } else if (WIFSIGNALED(status)) {
-    printf("inotify: %s exited with status %d (signal: %d)\n", callback[0], status, WTERMSIG(status));
+    printf("inotify: %s exited with status %d (signal: %d)\n",
+           callback[0], status, WTERMSIG(status));
   } else {
     printf("inotify: %s exited with status %d\n", callback[0], status);
   }
@@ -70,13 +74,14 @@ void run_callback(char** callback, struct timespec *last, bool always) {
       exit(EXIT_FAILURE);
     }
 
-    long diff_milliseconds = (tp.tv_sec - last->tv_sec) * 1000 \
+    int diff_milliseconds = (tp.tv_sec - last->tv_sec) * 1000 \
                    + (tp.tv_nsec - last->tv_nsec) / 1000000;
     if (diff_milliseconds < kPollTimeout) {
       return;
     }
 
-    printf("inotify: calling %s after %ldms since the last run\n", callback[0], diff_milliseconds);
+    printf("inotify: calling %s after %dms since the last run\n",
+           callback[0], diff_milliseconds);
   }
 
   run_callback_internal(callback);
@@ -91,10 +96,14 @@ void run_callback(char** callback, struct timespec *last, bool always) {
 int main(int argc, char* argv[]) {
   if (argc < 4) {
     printf("Usage: %s path file callback [callback-arg]...\n", argv[0]);
-    printf("This utility watches inotify events at 'path', optionally filtered by file name 'file',\n"
-           "and calls 'callback' with 'callback-arg' upon inotify events firing. Additionally,\n"
-           "'callback' is also called at start up or every %d milliseconds if no inotify event fires.\n"
-           "This program exits as success when 'callback' exits as success, or keeps running otherwise.\n",
+    printf("This utility watches inotify events at 'path', "
+           "optionally filtered by file name 'file',\n"
+           "and calls 'callback' with 'callback-arg' "
+           "upon inotify events firing. Additionally,\n"
+           "'callback' is also called at start up or "
+           "every %d milliseconds if no inotify event fires.\n"
+           "This program exits as success when 'callback' "
+           "exits as success, or keeps running otherwise.\n",
            kPollTimeout);
     exit(EXIT_FAILURE);
   }
@@ -174,7 +183,8 @@ int main(int argc, char* argv[]) {
       continue;
     }
 
-    printf("inotify: calling %s for %d matching event(s)\n", callback[0], matches);
+    printf("inotify: calling %s for %d matching event(s)\n",
+           callback[0], matches);
     run_callback(callback, &tp, true);
   }
 }

--- a/scripts/install-cni.sh
+++ b/scripts/install-cni.sh
@@ -86,21 +86,25 @@ fi
 
 if [ -f "/host/home/kubernetes/bin/gke" ]
 then
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniType#gke#g")
 else
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniType#ptp#g")
 fi
 
 if [ "${ENABLE_BANDWIDTH_PLUGIN}" == "true" ] && [ -f "/host/home/kubernetes/bin/bandwidth" ]
 then
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniBandwidthPlugin#,{\"type\": \"bandwidth\",\"capabilities\": {\"bandwidth\": true}}#g")
 else
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniBandwidthPlugin##g")
 fi
 
 token=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
 node_url="https://${KUBERNETES_SERVICE_HOST}:${KUBERNETES_SERVICE_PORT}/api/v1/nodes/${HOSTNAME}"
-response=$(curl -k -s -H "Authorization: Bearer $token" $node_url)
+response=$(curl -k -s -H "Authorization: Bearer $token" "$node_url")
 
 if [ "${MIGRATE_TO_DPV2:-}" == "true" ]
 then
@@ -115,16 +119,19 @@ fi
 if [ "${ENABLE_CILIUM_PLUGIN}" == "true" ]
 then
   echo "Adding Cilium plug-in to the CNI config."
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniCiliumPlugin#,{\"type\": \"cilium-cni\"}#g")
   # inotify calls back to the beginning of this script.
   inotify /host/home/kubernetes/bin cilium-cni "$0" cilium_ready
   echo "Cilium plug-in now confirmed as ready."
 else
   echo "Not using Cilium plug-in."
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@cniCiliumPlugin##g")
 fi
 
 # Fill CNI spec template.
+# shellcheck disable=SC2086
 ipv4_subnet=$(echo $response | jq '.spec.podCIDR')
 
 if [[ "${ipv4_subnet:-}" =~ ^\"[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*/[0-9][0-9]*\"$ ]]; then
@@ -137,6 +144,7 @@ else
 fi
 
 echo "Filling IPv4 subnet ${ipv4_subnet:-}"
+# shellcheck disable=SC2001,SC2086
 cni_spec=$(echo ${cni_spec:-} | sed -e "s#@ipv4Subnet#[{\"subnet\": ${ipv4_subnet:-}}]#g")
 
 if [ "$ENABLE_MASQUERADE" == "true" ]; then
@@ -176,6 +184,7 @@ if [ "${ENABLE_PRIVATE_IPV6_ACCESS:-}" == "true" ] || [ "$ENABLE_IPV6" == "true"
   if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
     echo "Found nic0 IPv6 address ${node_ipv6_addr:-}. Filling IPv6 subnet and route..."
 
+    # shellcheck disable=SC2001,SC2086
     cni_spec=$(echo ${cni_spec:-} | sed -e \
       "s#@ipv6SubnetOptional#, [{\"subnet\": \"${node_ipv6_addr:-}/112\"}]#g;
        s#@ipv6RouteOptional#, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}#g")
@@ -207,11 +216,13 @@ if [ "${ENABLE_PRIVATE_IPV6_ACCESS:-}" == "true" ] || [ "$ENABLE_IPV6" == "true"
     fi
   else
     echo "No IPv6 address found for nic0. Clearing IPv6 subnet and route..."
+    # shellcheck disable=SC2001,SC2086
     cni_spec=$(echo ${cni_spec:-} | \
       sed -e "s#@ipv6SubnetOptional##g; s#@ipv6RouteOptional##g")
   fi
 else
   echo "Clearing IPv6 subnet and route given private IPv6 access is disabled..."
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | \
     sed -e "s#@ipv6SubnetOptional##g; s#@ipv6RouteOptional##g")
 fi
@@ -229,10 +240,12 @@ default_nic=$(route -n | grep -E '^0\.0\.0\.0\s+\S+\s+0\.0\.0\.0' | grep -oE '\S
 
 # Set mtu
 if [ -f "/sys/class/net/$default_nic/mtu" ]; then
-  MTU=$(cat /sys/class/net/$default_nic/mtu)
+  MTU=$(cat "/sys/class/net/$default_nic/mtu")
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@mtu#$MTU#g")
   echo "Set the default mtu to $MTU, inherited from dev $default_nic"
 else
+  # shellcheck disable=SC2001,SC2086
   cni_spec=$(echo ${cni_spec:-} | sed -e "s#@mtu#1460#g")
   echo "Failed to read mtu from dev $default_nic, set the default mtu to 1460"
 fi
@@ -248,13 +261,14 @@ else
 fi
 
 # Atomically write CNI spec
-temp_file=$(mktemp ${output_file}.tmp.XXXXXX)
+temp_file=$(mktemp "${output_file}.tmp.XXXXXX")
+# shellcheck disable=SC2181
 if [ $? -ne 0 ]; then
   echo "Failed to create temp file, Exiting (1)..."
   exit 1
 fi
-trap "rm -f ${temp_file}" EXIT
-cat > ${temp_file} <<EOF
+trap 'rm -f "${temp_file}"' EXIT
+cat > "${temp_file}" <<EOF
 ${cni_spec:-}
 EOF
-mv ${temp_file} ${output_file}
+mv "${temp_file}" "${output_file}"

